### ISSUE TITLE
VM doesn't boot correctly after restoring its boot disk

### ISSUE THREAD

**Issue by khaledkbadr** [zero-os/0-Disk#425]
**Steps to reproduce**:
* Copy vdisk from a template (vd2)
* Start a VM using qemu with nbd, tlog servers
* echo a file for in the VM after it boots
* Stop the VM (and stopping nbd, tlog servers)
* restore the vdisk `/bin/zeroctl restore vdisk -f vd2 --config 192.168.122.79:2379 --end-timestamp 1504018830000000000 --data-shards 1 -v --parity-shards 1`
* Start the vm again with nbd, tlog servers, it halts at this 
![image](https://user-images.githubusercontent.com/6059889/29828478-79159468-8ce5-11e7-859f-9a77f58a5a59.png)

**config in etcd**
```
vd2:vdisk:conf:static
blockSize: 4096
readOnly: false
size: 4
type: boot

vd2:vdisk:conf:storage:nbd
slaveStorageClusterID: ''
storageClusterID: myblockcluster
templateStorageClusterID: 87ec688e821f5f0006619b7207d656bd
tlogServerClusterID: tlogserver_vm0_525400123456

vd2:vdisk:conf:storage:tlog
zeroStorClusterID: myobjectcluster


myblockcluster:cluster:conf:storage
dataStorage:
- address: 192.168.122.76:2000
- address: 192.168.122.79:2000
label: myblockcluster
metadataStorage:
  address: 192.168.122.79:2001
nodes:
- '525400123456'
- '525400123459'
status: ready

myobjectcluster:cluster:conf:zerostor
iyo:
  clientID: <clientID>
  namespace: test
  org: zero-khaled
  secret: <secret>
metadataServers:
- address: 192.168.122.79:2379
servers:
- address: 192.168.122.76:2002
- address: 192.168.122.79:2002

vm0:nbdserver:conf:vdisks
vdisks:
- vd2

tlogserver_vm0_525400123456:cluster:conf:tlog
servers:
- 192.168.122.79:11211
```

**Comment by iwanbk** [comment]
Did you see something in the logs of nbdserver and tlogserver?

**Comment by khaledkbadr** [comment]
Nothing that raises red flags, just success negotiations....etc

**Comment by iwanbk** [comment]
@khaledkbadr need logs of these:
- nbd & tlog before restore : normal logs is enough
- restoring : verbose log
- nbd & tlog after restore : verbose log

**Comment by khaledkbadr** [comment]
**Logs of nbdserver before restore**
```
t=2017-08-30T06:06:50+0000 lvl=info msg="Starting listening on unix:/server.socket.vm0" module=nbdserver caller=listener.go:151
t=2017-08-30T06:06:51+0000 lvl=info msg="Connect to unix:/server.socket.vm0 from @" module=nbdserver caller=listener.go:165
t=2017-08-30T06:06:51+0000 lvl=info msg="Getting vdisk \"vd2\"" module=global caller=exportcontroller.go:52
t=2017-08-30T06:06:52+0000 lvl=info msg="Negotiation succeeded with @/vd2" module=nbdserver caller=connection.go:609
t=2017-08-30T06:11:13+0000 lvl=info msg="vdisk 'vd2' received `Stop` command from vdisk completion" module=global caller=backend.go:232
t=2017-08-30T06:11:28+0000 lvl=info msg="Client @/vd2 cannot read data to write: unexpected EOF" module=nbdserver caller=connection.go:398
t=2017-08-30T06:11:28+0000 lvl=info msg="Receiver exiting for @/vd2" module=nbdserver caller=connection.go:224
t=2017-08-30T06:11:28+0000 lvl=info msg="Worker forced close for @/vd2" module=nbdserver caller=connection.go:634
t=2017-08-30T06:11:28+0000 lvl=info msg="Client @/vd2 cannot write reply: write unix /server.socket.vm0->@: write: broken pipe" module=nbdserver caller=connection.go:205
t=2017-08-30T06:11:28+0000 lvl=info msg="Replyer exiting for @/vd2" module=nbdserver caller=connection.go:188
t=2017-08-30T06:11:28+0000 lvl=info msg="GoBackground exited for vdisk: vd2" module=global caller=tlog.go:294
t=2017-08-30T06:11:28+0000 lvl=info msg="error while closing tlog client: close tcp 192.168.122.76:34906->192.168.122.79:11211: use of closed network connection" module=global caller=tlog.go:297
t=2017-08-30T06:11:28+0000 lvl=info msg="tlogclient : reconnect from read because of : read tcp 192.168.122.76:34906->192.168.122.79:11211: use of closed network connection" module=global caller=client.go:334
t=2017-08-30T06:11:28+0000 lvl=info msg="tlogclient: reconnect because `read tcp 192.168.122.76:34906->192.168.122.79:11211: use of closed network connection` failed: client already closed" module=global caller=client.go:340
t=2017-08-30T06:11:33+0000 lvl=eror msg="tlogclient failed to send: write tcp 192.168.122.76:34906->192.168.122.79:11211: use of closed network connection" module=global caller=client.go:543
t=2017-08-30T06:11:33+0000 lvl=info msg="tlog client : reconnect from send attemp(1) failed:client already closed" module=global caller=client.go:534
t=2017-08-30T06:11:38+0000 lvl=eror msg="tlogclient failed to send: write tcp 192.168.122.76:34906->192.168.122.79:11211: use of closed network connection" module=global caller=client.go:543
t=2017-08-30T06:11:38+0000 lvl=info msg="tlog client : reconnect from send attemp(1) failed:client already closed" module=global caller=client.go:534
t=2017-08-30T06:11:43+0000 lvl=eror msg="tlogclient failed to send: write tcp 192.168.122.76:34906->192.168.122.79:11211: use of closed network connection" module=global caller=client.go:543
t=2017-08-30T06:11:43+0000 lvl=info msg="tlog client : reconnect from send attemp(1) failed:client already closed" module=global caller=client.go:534
t=2017-08-30T06:11:43+0000 lvl=info msg="vdisk 'vd2' finished the flush under SIGTERM handler" module=global caller=backend.go:245
t=2017-08-30T06:11:43+0000 lvl=info msg="Backend background thread exiting for @/vd2" module=nbdserver caller=connection.go:616
t=2017-08-30T06:11:43+0000 lvl=info msg="Closed connection from @/vd2" module=nbdserver caller=connection.go:596
flush timeout
```
**Logs of tlog before restore** (The second log due to shutting down of the nbdserver I think)
```t=2017-08-30T06:06:51+0000 lvl=info msg="vdisk vd2 reset first sequence to 0" module=global caller=vdisk.go:232
t=2017-08-30T06:11:28+0000 lvl=info msg="failed to send resp to :vd2, err:writev tcp 192.168.122.79:11211->192.168.122.76:34906: writev: connection reset by peer" module=global caller=server.go:302
```

**Comment by khaledkbadr** [comment]
**zeroctl restore logs** (this is a short version of the same logs basically)
```
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
flushing persistent storage" module=global caller=storage.go:93 
```

**nbd logs after restore**
```
t=2017-08-30T06:29:07+0000 lvl=dbug msg="flags parsed: memorystorage=false tlsonly=false profileaddress=\"\" protocol=\"unix\" address=\"/server.socket.vm0\" config=\"192.168.122.79:2379\" lbacachelimit=20971520 logfile=\"\" id=\"vm0\"" module=global caller=main.go:73
t=2017-08-30T06:29:07+0000 lvl=info msg="Starting listening on unix:/server.socket.vm0" module=nbdserver caller=listener.go:151
t=2017-08-30T06:29:07+0000 lvl=dbug msg="started ExportController background thread (config hot reloading)" module=global caller=exportcontroller.go:102
t=2017-08-30T06:29:07+0000 lvl=dbug msg="watch goroutine for etcd key 'vm0:nbdserver:conf:vdisks' started" module=global caller=source_etcd.go:86
t=2017-08-30T06:29:07+0000 lvl=dbug msg="WatchNBDVdisksConfig for server vm0 started" module=global caller=config_api.go:187
t=2017-08-30T06:29:07+0000 lvl=info msg="Connect to unix:/server.socket.vm0 from @" module=nbdserver caller=listener.go:165
t=2017-08-30T06:29:07+0000 lvl=dbug msg="Start negotation with@" module=nbdserver caller=connection.go:599
t=2017-08-30T06:29:07+0000 lvl=dbug msg="Sending newstyle header to@" module=nbdserver caller=connection.go:644
t=2017-08-30T06:29:07+0000 lvl=dbug msg="Receiving client flags from@" module=nbdserver caller=connection.go:661
t=2017-08-30T06:29:07+0000 lvl=dbug msg="Receiving options from@" module=nbdserver caller=connection.go:670
t=2017-08-30T06:29:07+0000 lvl=dbug msg="Received option 3 from @" module=nbdserver caller=connection.go:686
t=2017-08-30T06:29:07+0000 lvl=dbug msg="Received option 1 from @" module=nbdserver caller=connection.go:686
t=2017-08-30T06:29:07+0000 lvl=dbug msg="getting exportConfig for: vd2" module=nbdserver caller=connection.go:751
t=2017-08-30T06:29:07+0000 lvl=info msg="Getting vdisk \"vd2\"" module=global caller=exportcontroller.go:52
t=2017-08-30T06:29:07+0000 lvl=dbug msg="generating export using driver ardb for @" module=nbdserver caller=connection.go:1065
t=2017-08-30T06:29:07+0000 lvl=dbug msg="create nbd config listener for vd2" module=global caller=provider.go:183
t=2017-08-30T06:29:08+0000 lvl=dbug msg="nbdStorageConfigWatcher (vd2) loading primary storage cluster: myblockcluster" module=global caller=composition_api.go:724
t=2017-08-30T06:29:08+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk vd2 started" module=global caller=config_api.go:253
t=2017-08-30T06:29:08+0000 lvl=dbug msg="watch goroutine for etcd key 'vd2:vdisk:conf:storage:nbd' started" module=global caller=source_etcd.go:86
t=2017-08-30T06:29:08+0000 lvl=dbug msg="nbdStorageConfigWatcher (vd2) loading template storage cluster: 87ec688e821f5f0006619b7207d656bd" module=global caller=composition_api.go:732
t=2017-08-30T06:29:08+0000 lvl=dbug msg="WatchStorageClusterConfig for cluster myblockcluster started" module=global caller=config_api.go:385
t=2017-08-30T06:29:08+0000 lvl=dbug msg="watch goroutine for etcd key 'myblockcluster:cluster:conf:storage' started" module=global caller=source_etcd.go:86
t=2017-08-30T06:29:08+0000 lvl=dbug msg="nbdStorageConfigWatcher (vd2) loading slave storage cluster: " module=global caller=composition_api.go:741
t=2017-08-30T06:29:08+0000 lvl=dbug msg="nbdStorageConfigWatcher (vd2) already watches slave cluster: " module=global caller=composition_api.go:864
t=2017-08-30T06:29:08+0000 lvl=dbug msg="spawn redisProvider listener goroutine for vd2" module=global caller=provider.go:194
t=2017-08-30T06:29:08+0000 lvl=dbug msg="creating LBA for vdisk vd2 with 64 bucket(s)" module=global caller=lba.go:47
t=2017-08-30T06:29:08+0000 lvl=dbug msg="WatchStorageClusterConfig for cluster 87ec688e821f5f0006619b7207d656bd started" module=global caller=config_api.go:385
t=2017-08-30T06:29:08+0000 lvl=dbug msg="watch goroutine for etcd key '87ec688e821f5f0006619b7207d656bd:cluster:conf:storage' started" module=global caller=source_etcd.go:86
t=2017-08-30T06:29:08+0000 lvl=dbug msg="creating tlogStorage for backend vd2 (boot)" module=global caller=backend_factory.go:114
t=2017-08-30T06:29:08+0000 lvl=dbug msg="watch goroutine for etcd key 'tlogserver_vm0_525400123456:cluster:conf:tlog' started" module=global caller=source_etcd.go:86
t=2017-08-30T06:29:08+0000 lvl=dbug msg="WatchTlogClusterConfig for cluster tlogserver_vm0_525400123456 started" module=global caller=config_api.go:516
t=2017-08-30T06:29:08+0000 lvl=dbug msg="generating backend ardb, using 333 flags, for @" module=nbdserver caller=connection.go:1120
t=2017-08-30T06:29:08+0000 lvl=info msg="Negotiation succeeded with @/vd2" module=nbdserver caller=connection.go:609
t=2017-08-30T06:29:08+0000 lvl=dbug msg="Start transmition phase with@/vd2" module=nbdserver caller=connection.go:624
t=2017-08-30T06:29:08+0000 lvl=dbug msg="background goroutine spawned for tlogstorage vd2" module=global caller=tlog.go:291
t=2017-08-30T06:29:08+0000 lvl=dbug msg="starting background thread for vdisk vd2's backend" module=global caller=backend.go:218
```
**tlog logs after restore**
```
t=2017-08-30T06:29:07+0000 lvl=dbug msg="flags parsed: memorystorage=false tlsonly=false profileaddress=\"\" protocol=\"unix\" address=\"/server.socket.vm0\" config=\"192.168.122.79:2379\" lbacachelimit=20971520 logfile=\"\" id=\"vm0\"" module=global caller=main.go:73
t=2017-08-30T06:29:07+0000 lvl=info msg="Starting listening on unix:/server.socket.vm0" module=nbdserver caller=listener.go:151
t=2017-08-30T06:29:07+0000 lvl=dbug msg="started ExportController background thread (config hot reloading)" module=global caller=exportcontroller.go:102
t=2017-08-30T06:29:07+0000 lvl=dbug msg="watch goroutine for etcd key 'vm0:nbdserver:conf:vdisks' started" module=global caller=source_etcd.go:86
t=2017-08-30T06:29:07+0000 lvl=dbug msg="WatchNBDVdisksConfig for server vm0 started" module=global caller=config_api.go:187
t=2017-08-30T06:29:07+0000 lvl=info msg="Connect to unix:/server.socket.vm0 from @" module=nbdserver caller=listener.go:165
t=2017-08-30T06:29:07+0000 lvl=dbug msg="Start negotation with@" module=nbdserver caller=connection.go:599
t=2017-08-30T06:29:07+0000 lvl=dbug msg="Sending newstyle header to@" module=nbdserver caller=connection.go:644
t=2017-08-30T06:29:07+0000 lvl=dbug msg="Receiving client flags from@" module=nbdserver caller=connection.go:661
t=2017-08-30T06:29:07+0000 lvl=dbug msg="Receiving options from@" module=nbdserver caller=connection.go:670
t=2017-08-30T06:29:07+0000 lvl=dbug msg="Received option 3 from @" module=nbdserver caller=connection.go:686
t=2017-08-30T06:29:07+0000 lvl=dbug msg="Received option 1 from @" module=nbdserver caller=connection.go:686
t=2017-08-30T06:29:07+0000 lvl=dbug msg="getting exportConfig for: vd2" module=nbdserver caller=connection.go:751
t=2017-08-30T06:29:07+0000 lvl=info msg="Getting vdisk \"vd2\"" module=global caller=exportcontroller.go:52
t=2017-08-30T06:29:07+0000 lvl=dbug msg="generating export using driver ardb for @" module=nbdserver caller=connection.go:1065
t=2017-08-30T06:29:07+0000 lvl=dbug msg="create nbd config listener for vd2" module=global caller=provider.go:183
t=2017-08-30T06:29:08+0000 lvl=dbug msg="nbdStorageConfigWatcher (vd2) loading primary storage cluster: myblockcluster" module=global caller=composition_api.go:724
t=2017-08-30T06:29:08+0000 lvl=dbug msg="watchVdiskNBDConfig for vdisk vd2 started" module=global caller=config_api.go:253
t=2017-08-30T06:29:08+0000 lvl=dbug msg="watch goroutine for etcd key 'vd2:vdisk:conf:storage:nbd' started" module=global caller=source_etcd.go:86
t=2017-08-30T06:29:08+0000 lvl=dbug msg="nbdStorageConfigWatcher (vd2) loading template storage cluster: 87ec688e821f5f0006619b7207d656bd" module=global caller=composition_api.go:732
t=2017-08-30T06:29:08+0000 lvl=dbug msg="WatchStorageClusterConfig for cluster myblockcluster started" module=global caller=config_api.go:385
t=2017-08-30T06:29:08+0000 lvl=dbug msg="watch goroutine for etcd key 'myblockcluster:cluster:conf:storage' started" module=global caller=source_etcd.go:86
t=2017-08-30T06:29:08+0000 lvl=dbug msg="nbdStorageConfigWatcher (vd2) loading slave storage cluster: " module=global caller=composition_api.go:741
t=2017-08-30T06:29:08+0000 lvl=dbug msg="nbdStorageConfigWatcher (vd2) already watches slave cluster: " module=global caller=composition_api.go:864
t=2017-08-30T06:29:08+0000 lvl=dbug msg="spawn redisProvider listener goroutine for vd2" module=global caller=provider.go:194
t=2017-08-30T06:29:08+0000 lvl=dbug msg="creating LBA for vdisk vd2 with 64 bucket(s)" module=global caller=lba.go:47
t=2017-08-30T06:29:08+0000 lvl=dbug msg="WatchStorageClusterConfig for cluster 87ec688e821f5f0006619b7207d656bd started" module=global caller=config_api.go:385
t=2017-08-30T06:29:08+0000 lvl=dbug msg="watch goroutine for etcd key '87ec688e821f5f0006619b7207d656bd:cluster:conf:storage' started" module=global caller=source_etcd.go:86
t=2017-08-30T06:29:08+0000 lvl=dbug msg="creating tlogStorage for backend vd2 (boot)" module=global caller=backend_factory.go:114
t=2017-08-30T06:29:08+0000 lvl=dbug msg="watch goroutine for etcd key 'tlogserver_vm0_525400123456:cluster:conf:tlog' started" module=global caller=source_etcd.go:86
t=2017-08-30T06:29:08+0000 lvl=dbug msg="WatchTlogClusterConfig for cluster tlogserver_vm0_525400123456 started" module=global caller=config_api.go:516
t=2017-08-30T06:29:08+0000 lvl=dbug msg="generating backend ardb, using 333 flags, for @" module=nbdserver caller=connection.go:1120
t=2017-08-30T06:29:08+0000 lvl=info msg="Negotiation succeeded with @/vd2" module=nbdserver caller=connection.go:609
t=2017-08-30T06:29:08+0000 lvl=dbug msg="Start transmition phase with@/vd2" module=nbdserver caller=connection.go:624
t=2017-08-30T06:29:08+0000 lvl=dbug msg="background goroutine spawned for tlogstorage vd2" module=global caller=tlog.go:291
t=2017-08-30T06:29:08+0000 lvl=dbug msg="starting background thread for vdisk vd2's backend" module=global caller=backend.go:218
```

**Comment by iwanbk** [comment]
> t=2017-08-30T06:11:28+0000 lvl=info msg="GoBackground exited for vdisk: vd2" module=global caller=tlog.go:294
> t=2017-08-30T06:11:28+0000 lvl=info msg="error while closing tlog client: close tcp 192.168.122.76:34906->192.168.122.79:11211: use of closed network connection" module=global caller=tlog.go:297
> t=2017-08-30T06:11:28+0000 lvl=info msg="tlogclient : reconnect from read because of : read tcp 192.168.122.76:34906->192.168.122.79:11211: use of closed network connection" module=global caller=client.go:334
> t=2017-08-30T06:11:28+0000 lvl=info msg="tlogclient: reconnect because `read tcp 192.168.122.76:34906->192.168.122.79:11211: use of closed network connection` failed: client already closed" module=global caller=client.go:340
> t=2017-08-30T06:11:33+0000 lvl=eror msg="tlogclient failed to send: write tcp 192.168.122.76:34906->192.168.122.79:11211: use of closed network connection" module=global caller=client.go:543
> t=2017-08-30T06:11:33+0000 lvl=info msg="tlog client : reconnect from send attemp(1) failed:client already closed" module=global caller=client.go:534
> t=2017-08-30T06:11:38+0000 lvl=eror msg="tlogclient failed to send: write tcp 192.168.122.76:34906->192.168.122.79:11211: use of closed network connection" module=global caller=client.go:543
> t=2017-08-30T06:11:38+0000 lvl=info msg="tlog client : reconnect from send attemp(1) failed:client already closed" module=global caller=client.go:534
> t=2017-08-30T06:11:43+0000 lvl=eror msg="tlogclient failed to send: write tcp 192.168.122.76:34906->192.168.122.79:11211: use of closed network connection" module=global caller=client.go:543
> t=2017-08-30T06:11:43+0000 lvl=info msg="tlog client : reconnect from send attemp(1) failed:client already closed" module=global caller=client.go:534
> t=2017-08-30T06:11:43+0000 lvl=info msg="vdisk 'vd2' finished the flush under SIGTERM handler" module=global caller=backend.go:245
> t=2017-08-30T06:11:43+0000 lvl=info msg="Backend background thread exiting for @/vd2" module=nbdserver caller=connection.go:616

Thanks @khaledkbadr , there is no problem from the logs. It is tlogclient that didn't disconnect in gracefull way, addressed at #426

**Comment by GlenDC** [comment]
> there is no problem from the logs. It is tlogclient that didn't disconnect in gracefull way

Is still simply something you observe from the logs, or does this also explain @khaledkbadr his issue?

**Comment by iwanbk** [comment]
> Is still simply something you observe from the logs, or does this also explain @khaledkbadr his issue?

khaled told me in the chat that he see problem in `logs before restore`, i told him that it is not a problem, and will  explain to  him here :)

**Comment by GlenDC** [comment]
Ah ok. So that means the issue origin is still unknown :(

**Comment by iwanbk** [comment]
> tlog logs after restore

It seems you provide nbd instead of tlog logs

**Comment by khaledkbadr** [comment]
I did, but edited. refresh :P

**Comment by iwanbk** [comment]
@GlenDC do you see some problems there?

**Comment by GlenDC** [comment]
No, nbdserver logs doesn't seem to be verbose-enabled though. But I can't see anything wrong.

I do know that the restore tlog code could use more verbose logs, because right now the `zeroctl restore vdisk -v` really doesn't give any verbose logging at all, besides the flush messages you see in the logs above. Perhaps that could help with debugging this issue.

@khaledkbadr is this issue only with a specific image, or is it with any image? if it is with a specific, you could perhaps give it to @iwanbk, as that might help him with reproducing it on his dev machine.

**Comment by khaledkbadr** [comment]
We use the ubuntu-1604 cloud image (hacked version of it so we can login), You can find it in hub.gig.com under `template:ubuntu-1604`

**Comment by GlenDC** [comment]
@khaledkbadr hmm weird, I use a very similar image, among other images I test with. Although the ubuntu-1604 cloud image I use is the original one as downloaded. Yet I can restore with it... At least when I tried last.

**Comment by khaledkbadr** [comment]
hmmmmm, All I know is that the issue appeared after we integrated 0-stor with tlogserver. And it seems to be persistent issue not just now and then kinda issue.

**Comment by iwanbk** [comment]
> hmmmmm, All I know is that the issue appeared after we integrated 0-stor with tlogserver. And it seems to be persistent issue not just now and then kinda issue.

Confusing, i just redo the test yesterday:
- start vm
- destroy nbd's ardb 
- restore
- start vm again

And all was good.
Maybe i need different procedures...

**Comment by khaledkbadr** [comment]
we don't destroy the ardb, just stop the nbdserver, tlogserver

**Comment by iwanbk** [comment]
> we don't destroy the ardb, just stop the nbdserver, tlogserver

Why you need to restore when you still have the ardb?

**Comment by khaledkbadr** [comment]
we use the restore for rollback of vdisk not to restore after delete.

**Comment by GlenDC** [comment]
@iwanbk restore command is also used to do rollbacks. In which case the vdisk gets deleted first, and than the vdisk gets rebuild using the restore command.

**Comment by iwanbk** [comment]
> we use the restore for rollback of vdisk not to restore after delete.

Could you also try the `restore` after delete, not  `restore` for `rollback`?

> @iwanbk restore command is also used to do rollbacks.

Yeah, i know. I was expecting `rollback` action to use `rollback` word instead of `restore`.

**Comment by khaledkbadr** [comment]
We use the `-f` flag, that does the same thing. no?

**Comment by iwanbk** [comment]
> We use the -f flag, that does the same thing. no?

Yeah, it should. I'll test

**Comment by iwanbk** [comment]
-f flag  worked for me,

**Comment by GlenDC** [comment]
Yes, defining the `-f` flag in the `zeroctl restore vdisk` command should at all times be the same logic as a combined `zeroctl delete vdisk && zeroctl restore vdisk` command. If that's not the case that would be a bug. In the current version I'm sure they are the same, but mentioning it for future encounters, should you ever have suspicions.

**Comment by iwanbk** [comment]
```
t=2017-08-30T06:11:28+0000 lvl=info msg="Client @/vd2 cannot read data to write: unexpected EOF" module=nbdserver caller=connection.go:398
t=2017-08-30T06:11:28+0000 lvl=info msg="Receiver exiting for @/vd2" module=nbdserver caller=connection.go:224
t=2017-08-30T06:11:28+0000 lvl=info msg="Worker forced close for @/vd2" module=nbdserver caller=connection.go:634
t=2017-08-30T06:11:28+0000 lvl=info msg="Client @/vd2 cannot write reply: write unix /server.socket.vm0->@: write: broken pipe" module=nbdserver caller=connection.go:205
t=2017-08-30T06:11:28+0000 lvl=info msg="Replyer exiting for @/vd2" module=nbdserver caller=connection.go:188
```

I was too focused with tlog so didn't notice these lines.
There is some error in nbdserver module.
The error seems to be between nbd client & server.

> Stop the VM (and stopping nbd, tlog servers)

@khaledkbadr can you make sure that you stop nbd & tlog first before stop the VM?

**Comment by iwanbk** [comment]
> @khaledkbadr can you make sure that you stop nbd & tlog first before stop the VM?

My bad, i forgot that qemu will automatically wait for all data to be flushed, right?

**Comment by khaledkbadr** [comment]
> Which means the VM wait for nbd & tlog to do flush & stop.
I guess not I test on the `stopping` the vm not `shutting it down`, cause we need the rollback most in stopping the vm non-gracefully since data loss may happen due to it. That's why we asked for a flush on SIGTERM signal for the nbdserver.

**Comment by iwanbk** [comment]
> I guess not I test on the stopping the vm not shutting it down, cause we need the rollback most in stopping the vm non-gracefully since data loss may happen due to it. That's why we asked for a flush on SIGTERM signal for the nbdserver.

by `stopping`, you mean the same case as #288, right?
It actually wait for some `reasonable timeout`, then destroy the VM.
But yes, it doesn't do graceful shutdown, only waiting for `reasonable timeout`
See https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainDestroy

From the ndb logs, SIGTERM still handled properly.

From the gondbserver error logs above (especially the line below), it seems that the flush still not finished after the `reasonable timeout` and then qemu simply close the nbd connection.
**And there might be some data that not sent yet to nbd server.** 
```
=2017-08-30T06:11:28+0000 lvl=info msg="Client @/vd2 cannot read data to write: unexpected EOF" module=nbdserver caller=connection.go:398
```

You said that it started to happen after we were moving to 0-stor.
Until yesterday, i was pretty sure that 0-stor is way slower (which might causing unfinished flush).
But yesterday 0-stor server got quite performance improvement (https://github.com/zero-os/0-stor/pull/212).
Please make sure you have that improvement and redo the test.

**Comment by iwanbk** [comment]
> Copy vdisk from a template (vd2)
I also see some possible error here.

@GlenDC does the `zeroctl copy vdisk` also copy tlog data? I guess not, because it doesn't implemented yet, right? see #280 

@khaledkbadr did you also provide tlog related config when copying the  disk?

**Comment by khaledkbadr** [comment]
>@khaledkbadr did you also provide tlog related config when copying the disk?

This all I got in my etcd cluster + templatecluster config (which I forgot to add) but I guess it doesn't matter right?

> Until yesterday, i was pretty sure that 0-stor is way slower (which might causing unfinished flush).

Sure, I will try again after checking if my 0-stor flist is at the latest or not. But that may raise problems in the testing/production env since it uses much networking unlike my env which is totally local and still get this problem.

**Comment by iwanbk** [comment]
> This all I got in my etcd cluster + templatecluster config (which I forgot to add) but I guess it doesn't matter right?

I think it does matter, not sure, i don't really know what current `zeroctl copy` does. I'll check it.

So, if `zeroctl copy` doesn't write to tlog, then there is no way for tlog to restore it. Because tlog doesn't have the data it needs to restore.
If possible, please do the test without `zeroctl copy` step. I assume that #288 is without `zeroctl copy`, no?

> But that may raise problems in the testing/production env since it uses much networking unlike my env which is totally local and still get this problem.

Yeah, we are working to improve the 0-stor performance. https://github.com/zero-os/0-stor/pull/212 is one of it.

**Comment by khaledkbadr** [comment]
> the `zeroctl copy vdisk` step is only done at the very first when trying to copy `template data` to a disk ie `vd2`. Either than that it's not used at all. Actually if I skipped it the `vd2` would not have any data and it will not boot at all, so it can't be skipped

**Comment by iwanbk** [comment]
OK, i just checked the `zeroctl copy` command. I think it is not copying tlog data, as expected because we still have #280 open.

My conclusion so far, the problem could be caused by:

- slow 0-stor server, as mentioned above, some improvement at https://github.com/zero-os/0-stor/pull/212

- `zeroctl copy` doesn't copy tlog data. So, tlog doesn't have data which needed to restore the base ubuntu image. All tlog has is the transaction after VM booted up. In @khaledkbadr test is only single echo to a file
  But i'm still not sure if it is the case, because from the logs above, the flush need 40 seconds to finish, it must be a lot of transaction, not only single echo to a file.

   @GlenDC might have some explanation about this.

**Comment by GlenDC** [comment]
@iwanbk stop focussing on `zeroctl copy`. @khaledkbadr is using that command right at the start, before anything else, to copy the deduped metadata from a template vdisk, so that when he starts vd2 he can actually fetch that vdisk's data on the fly.

It's obviously true that `zeroctl copy` doesn't copy tlog (meta)data, hence why there is an issue open for it. And why we even talked about it yesterday or so.

So far the issue seems to be that the tlogserver shuts down before it is finished. Which is not unsurprising that it can happen. This is the main reason https://github.com/zero-os/0-Disk/issues/322 would be really useful. Because just waiting "reasonable" time from the nbdserver is flawed and not reasonable at all. Better it would be to have tlogserver _also_ handle SIGTERM so that it can shutdown _only_ when it is finished with its work. (I suppose you could put like a very big upper limit of 5 minutes (really not sure what a reasonable upper limit is here) or more, in case there is a bug in the tlogserver that causes it to not finish ever).

It's our best lead so far, as an unfinished tlogserver would definitely explain why the restored vdisk is not working as expected.

**Comment by iwanbk** [comment]
> So far the issue seems to be that the tlogserver shuts down before it is finished.

No, the log clearly mentioned that flush has been finished.

> Because just waiting "reasonable" time from the nbdserver is flawed and not reasonable at all

It is qemu that doing this, not tlog
You really missed the flow here

> Better it would be to have tlogserver also handle SIGTERM so that it can shutdown only when it is finished with its work

It is already finished it's work.

I need to repeat it again : 
- nbd flush handle sigterm
- nbd flush only finished after tlog finished it's work. In this point, tlog already finished it's work
- tlog get sigterm, no problem : it already has nothing to do.

**Comment by iwanbk** [comment]
> slow 0-stor server, as mentioned above, some improvement at zero-os/0-stor#212

It couldn't be the reason.
If this is the reason, then ndb will log 'flush timed out' instead of `finished the flush under SIGTERM handler`

**Comment by GlenDC** [comment]
Ok if you're so confident about that flow @iwanbk, than I'm out of suggestions for now, yet this bug is still there...

**Comment by khaledkbadr** [comment]
@iwanbk I've just tested with the latest of `0-stor` and I still got the same problem

**Comment by GlenDC** [comment]
@khaledkbadr you also have the latest of 0-Disk's `tlogserver` (master)?

**Comment by khaledkbadr** [comment]
Nope, I didn't update that. anything changed there?

**Comment by GlenDC** [comment]
I don't know at which version you were building before, so hard to tell for me if something changed. All changes can be found @ https://github.com/zero-os/0-Disk/commits/master, should give you the answer you're looking for. Just best to give all relevant build commits hashes each time you open an issue, and when you give comments where you updated one of them. This prevents confusion related to the version you might be on.

**Comment by khaledkbadr** [comment]
I just tried with both 0-stor and 0-disk master branches and still got the same problem. 
0-disk commit: ac3c10967ee69349581eee8cbf536e7add62798c
0-stor commit: ca6146cfada48dfb1e436fd5e3260912047001a2

**Comment by iwanbk** [comment]
```
t=2017-08-30T06:11:28+0000 lvl=info msg="Client @/vd2 cannot read data to write: unexpected EOF" module=nbdserver caller=connection.go:398
t=2017-08-30T06:11:28+0000 lvl=info msg="Receiver exiting for @/vd2" module=nbdserver caller=connection.go:224
t=2017-08-30T06:11:28+0000 lvl=info msg="Worker forced close for @/vd2" module=nbdserver caller=connection.go:634
t=2017-08-30T06:11:28+0000 lvl=info msg="Client @/vd2 cannot write reply: write unix /server.socket.vm0->@: write: broken pipe" module=nbdserver caller=connection.go:205
t=2017-08-30T06:11:28+0000 lvl=info msg="Replyer exiting for @/vd2" module=nbdserver caller=connection.go:188
```
@GlenDC any idea what happens here?

**Comment by GlenDC** [comment]
Seems to be just the end of a connection. It only says it's unexpected EOF because `ReadFull` (as done in gonbdserver) probably reports that like that.

However that info message is only returned if it can't read data from the open connection, after it has already received the CMD_WRITE command, so that does seem suspicious. But perhaps it's normal, I'm not too sure. Even so, that is an error that was logged before the restore process, and I think at that point the vdisk was still working as expected, so even it it were a problem, it wouldn't explain this (issue's) problem I think.

But it might also be an expected thing, as around that gonbdserver code there is also a check to see if it's not an error that we ourselves send, so it seems it might be expected that at that point a connection ends. But agian, I'm not too sure of this.

**Comment by iwanbk** [comment]
> Seems to be just the end of a connection. It only says it's unexpected EOF because ReadFull (as done in gonbdserver) probably reports that like that.

I didn't get this in my test.

I worry that it is because the qemu ungraceful shutdown

**Comment by GlenDC** [comment]
Your worries could be very quickly resolved in case @khaledkbadr could confirm that after `Logs of nbdserver before restore`, he can still boot up the vdisk or not.

**Comment by iwanbk** [comment]
> Your worries could be very quickly resolved in case @khaledkbadr could confirm that after Logs of nbdserver before restore, he can still boot up the vdisk or not.

Ah, yes, you're right. i think @khaledkbadr already did this and it worked

**Comment by iwanbk** [comment]
@khaledkbadr could you test with these branchs of 0-disk? 
- `nbd-eol-log` : it won't fix the issue, i only added some log there
- `nbd-eol-experiment` : (forked from nbd-eol-log) it might fix your issue or reveal other error(s). Even it fix your issue. it is not the proper solution, it is only a quick dirty hack

I need you to do these because:
- it is the part in nbd that i don't fully understand
- this part was changed

Thx

**Comment by khaledkbadr** [comment]
> nbd-eol-log


```
t=2017-09-05T08:08:16+0000 lvl=info msg="Starting listening on unix:/server.socket.vm0" module=nbdserver caller=listener.go:151
t=2017-09-05T08:08:16+0000 lvl=info msg="Connect to unix:/server.socket.vm0 from @" module=nbdserver caller=listener.go:165
t=2017-09-05T08:08:16+0000 lvl=info msg="Getting vdisk \"vd2\"" module=global caller=exportcontroller.go:52
t=2017-09-05T08:08:17+0000 lvl=info msg="Negotiation succeeded with @/vd2" module=nbdserver caller=connection.go:609
t=2017-09-05T08:08:19+0000 lvl=info msg="BACKEND FLUSH" module=global caller=backend.go:179
t=2017-09-05T08:08:19+0000 lvl=info msg="tlogStorage Flush()" module=global caller=tlog.go:211
t=2017-09-05T08:09:06+0000 lvl=info msg="BACKEND FLUSH" module=global caller=backend.go:179
t=2017-09-05T08:09:06+0000 lvl=info msg="tlogStorage Flush()" module=global caller=tlog.go:211
t=2017-09-05T08:09:06+0000 lvl=info msg="BACKEND FLUSH" module=global caller=backend.go:179
t=2017-09-05T08:09:06+0000 lvl=info msg="tlogStorage Flush()" module=global caller=tlog.go:211
t=2017-09-05T08:09:06+0000 lvl=info msg="BACKEND FLUSH" module=global caller=backend.go:179
t=2017-09-05T08:09:06+0000 lvl=info msg="tlogStorage Flush()" module=global caller=tlog.go:211
t=2017-09-05T08:09:06+0000 lvl=info msg="BACKEND FLUSH" module=global caller=backend.go:179
t=2017-09-05T08:09:06+0000 lvl=info msg="tlogStorage Flush()" module=global caller=tlog.go:211
t=2017-09-05T08:09:11+0000 lvl=info msg="BACKEND FLUSH" module=global caller=backend.go:179
t=2017-09-05T08:09:11+0000 lvl=info msg="tlogStorage Flush()" module=global caller=tlog.go:211
t=2017-09-05T08:09:13+0000 lvl=info msg="BACKEND FLUSH" module=global caller=backend.go:179
t=2017-09-05T08:09:13+0000 lvl=info msg="tlogStorage Flush()" module=global caller=tlog.go:211
t=2017-09-05T08:09:17+0000 lvl=info msg="BACKEND FLUSH" module=global caller=backend.go:179
t=2017-09-05T08:09:17+0000 lvl=info msg="tlogStorage Flush()" module=global caller=tlog.go:211
t=2017-09-05T08:09:17+0000 lvl=info msg="BACKEND FLUSH" module=global caller=backend.go:179
t=2017-09-05T08:09:17+0000 lvl=info msg="tlogStorage Flush()" module=global caller=tlog.go:211
```

**Comment by khaledkbadr** [comment]
i just tested with `nbd-eol-experiment` branch and still same behaviour after restore

**Comment by iwanbk** [comment]
> t=2017-09-05T08:18:47+0000 lvl=info msg="vdisk 'vd2' finished the flush under SIGTERM handler. err = flush timeout" module=global caller=backend.go:247

Thx, we can see that we got `flush timeout` here.
Current timeout in nbd is set to 20 seconds, which i think is too low but worked for ardb based tlog.

**Comment by iwanbk** [comment]
@khaledkbadr could you please try these branches:

- nbd-fix-close : it should fixes your issue, at least it fixes the issue on my machine. I could finally reproduce it.
- nbd-inc-tlog-timeout : not sure  it will fix your issue. I could reproduce it, but not with exact same flow as yours. It works on my machine but maybe not on yours

**Comment by khaledkbadr** [comment]
> tested with `nbd-fix-close` 

```
t=2017-09-05T10:01:43+0000 lvl=info msg="Starting listening on unix:/server.socket.vm0" module=nbdserver caller=listener.go:151
t=2017-09-05T10:01:43+0000 lvl=info msg="Connect to unix:/server.socket.vm0 from @" module=nbdserver caller=listener.go:165
t=2017-09-05T10:01:43+0000 lvl=info msg="Getting vdisk \"vd2\"" module=global caller=exportcontroller.go:52
t=2017-09-05T10:01:44+0000 lvl=info msg="Negotiation succeeded with @/vd2" module=nbdserver caller=connection.go:609
t=2017-09-05T10:05:50+0000 lvl=info msg="Client @/vd2 waiting for inflight requests prior to disconnect" module=nbdserver caller=connection.go:553
t=2017-09-05T10:05:50+0000 lvl=info msg="Client @/vd2 requested disconnect\n" module=nbdserver caller=connection.go:511
t=2017-09-05T10:05:50+0000 lvl=info msg="Receiver exiting for @/vd2" module=nbdserver caller=connection.go:224
t=2017-09-05T10:05:50+0000 lvl=info msg="Worker forced close for @/vd2" module=nbdserver caller=connection.go:634
t=2017-09-05T10:05:50+0000 lvl=info msg="tlogStorage close vdisk `vd2" module=global caller=tlog.go:257
t=2017-09-05T10:05:50+0000 lvl=info msg="GoBackground exited for vdisk: vd2" module=global caller=tlog.go:291
t=2017-09-05T10:05:50+0000 lvl=info msg="error while closing tlog client: close tcp 192.168.122.76:39180->192.168.122.79:11211: use of closed network connection" module=global caller=tlog.go:294
t=2017-09-05T10:05:50+0000 lvl=info msg="tlogclient : reconnect from read because of : read tcp 192.168.122.76:39180->192.168.122.79:11211: use of closed network connection" module=global caller=client.go:337
t=2017-09-05T10:05:50+0000 lvl=info msg="tlogclient: reconnect because `read tcp 192.168.122.76:39180->192.168.122.79:11211: use of closed network connection` failed: client already closed" module=global caller=client.go:343
t=2017-09-05T10:05:50+0000 lvl=info msg="Replyer exiting for @/vd2" module=nbdserver caller=connection.go:188
t=2017-09-05T10:05:50+0000 lvl=info msg="Backend background thread exiting for @/vd2" module=nbdserver caller=connection.go:616
t=2017-09-05T10:05:50+0000 lvl=info msg="Closed connection from @/vd2" module=nbdserver caller=connection.go:596

t=2017-09-05T10:08:26+0000 lvl=info msg="Starting listening on unix:/server.socket.vm0" module=nbdserver caller=listener.go:151
t=2017-09-05T10:08:27+0000 lvl=info msg="Connect to unix:/server.socket.vm0 from @" module=nbdserver caller=listener.go:165
t=2017-09-05T10:08:27+0000 lvl=info msg="Getting vdisk \"vd2\"" module=global caller=exportcontroller.go:52
t=2017-09-05T10:08:27+0000 lvl=info msg="Negotiation succeeded with @/vd2" module=nbdserver caller=connection.go:60
```

**Comment by iwanbk** [comment]
> tested with nbd-fix-close

The weird thing here is that i can't see any SIGTERM handler log. I might did very stupid mistake on my nbd.Close fix attempt

@khaledkbadr please try `nbd-eol-experiment` again.
That branch already gave us the expected logs, i increased the timeout now.

 (really sorry for this, i really need to reproduce this test on my own with 0-orch).

**Comment by khaledkbadr** [comment]
I still can't see the SIGTERM handler now too

![image](https://user-images.githubusercontent.com/6059889/30058706-cd4fb874-9244-11e7-851f-e309854d97ea.png)

**Comment by iwanbk** [comment]
> I still can't see the SIGTERM handler now too

i'm afraid that you did different test procedures, as all the update do is increase the timeout value

**Comment by khaledkbadr** [comment]
That's really weird, I just tried again and made sure I used the `nbd-eol-experiment` branch and got nearly the same logs

![image](https://user-images.githubusercontent.com/6059889/30059615-77d47368-9248-11e7-9081-9c2a9891153a.png)

**Comment by iwanbk** [comment]
> That's really weird, I just tried again and made sure I used the nbd-eol-experiment branch and got nearly the same logs

from `tlogStorage Close()` line in the log, you indeed seems to use the right branch.

But the nbd log must have this line to indicate that SIGTERM handler was working
```
"vdisk 'vd2' received `Stop` command from vdisk completion" module=global caller=backend.go:232
```

**Comment by khaledkbadr** [comment]
Since I use the Orchestrator which had the same code for all testing, I believe we still still send the `SIGTERM` signal. But I don't know why it doesn't react on it

**Comment by GlenDC** [comment]
Once this issue is resolved make sure to open a PR with the solution using [the 1.1.0-alpha-8 branch](https://github.com/zero-os/0-Disk/tree/1.1.0-alpha-7) as the base branch. Master is from now on meant for beta code.

**Comment by iwanbk** [comment]
@khaledkbadr 

If you are sure that `SIGTERM` already delivered, then it seems that nbd's vdisk worker indeed already stopped at the moment `SIGTERM` came (as we can see in the log, there is `tlogStorage Close()` there), which could be caused by:

- unexpected `Close` : `Close()` event that unexpected by nbd's tlogStorage to happen. The `Close()` handler doesn't have `Flush` code there, which might be needed in our case. Still checking it.

- expected `Close` : in this case, there is indeed no data to flush anymore, in other words tlog already fully synced with nbd. If this is what happened, the restore should succeed. please confirm that you already tried to restore and failed

**Comment by khaledkbadr** [comment]
I believe it's an expected `Close`, and I just confirmed that the restore results in the same behaviour.

**Comment by iwanbk** [comment]
> I believe it's an expected Close

Why you believe this?
the unexpected close didn't necessarily caused by testing procedures or testing code. It could be caused by the interaction between qemu & our nbd server.

**Comment by khaledkbadr** [comment]
Aha, I thought you meant the tlogserver closed, and even that the `tlogStorage Close()` was there before I stopped the vm (nbdserver, tlogserver). My bad :)

**Comment by iwanbk** [comment]
@khaledkbadr i updated `nbd-eol-experiment` branch, added some handler in case unexpected close happens. Could you please try it again?

**Comment by khaledkbadr** [comment]
Same behavior

![image](https://user-images.githubusercontent.com/6059889/30102500-8d3a4146-92f8-11e7-80ae-dd2c6ff8d468.png)

**Comment by khaledkbadr** [comment]
I tested with last version of the `nbd-eol-experiment` where Iwan added more logs
**nbd logs before restore**
![image](https://user-images.githubusercontent.com/6059889/30110178-d8ea1abc-9311-11e7-942d-4a9aa9e56fe4.png)

**zeroctl restore logs**
![image](https://user-images.githubusercontent.com/6059889/30110183-dbf68ee8-9311-11e7-9e9e-8d8ce8b96bc4.png)

**nbd logs after restore**
![image](https://user-images.githubusercontent.com/6059889/30110187-e0be22c4-9311-11e7-81bb-410747308d11.png)


Also @iwanbk noted that the number of blocks is low since I do 4k block size with 331M ubuntu image

**Comment by iwanbk** [comment]
> I tested with last version of the nbd-eol-experiment where Iwan added more logs

From the logs above:
- nbd closed with last sequence = 17693 and cache empty (or blocks already flushed)
- `zeroctl restore` also restore until 17693 blocks.

From above facts, i think tlog client, tlog server, and tlog replay (restore)  already do it's job.

> Also @iwanbk noted that the number of blocks is low since I do 4k block size with 331M ubuntu image

Yeah, i use smaller lede images and have 67K blocks.

The difference between my image and khaled's is that his image is copied image using `zeroctl copy` while mine is from `qemu-img convert`

My plan is checking how nbd's tlogStorage work with copied image.

**Comment by iwanbk** [comment]
> My plan is checking how nbd's tlogStorage work with copied image.

After checking, i am pretty sure that the problem lies in the copied image because it only copied the metadata.

It is true that when reading a disk, nbd also try write to primary cluster if it is not exist yet, which result in tlog transaction being sent to tlog server.

The problem is, booting a VM doesn't mean read the entire disk.Which make the tlog doesn't have all data it's need.

I'm working to provide quick fix as proof of concept to fix this issue.

**Comment by iwanbk** [comment]
> I'm working to provide quick fix as proof of concept to fix this issue.

The idea is, we also generate tlog transaction when copying vdisk.
We don't copy the tlog's data because not all template have it.
But we can improve it in the future.

**Comment by iwanbk** [comment]
@GlenDC 
I found that vdisk copy use raw redis connection instead of backend/storage object.
Which mean my plan won't work because it needs to get the block in order to generate tlog transaction.

Any idea for quick fix so we don't block the 0-orch team?

**Comment by GlenDC** [comment]
Are you 100% sure this is the issue? As generating tlog transactions during copy doesn't seem like a good idea to me anyhow.

**Comment by iwanbk** [comment]
> Are you 100% sure this is the issue? As generating tlog transactions during copy doesn't seem like a good idea to me anyhow.

Yes, please read more in telegram chat.
Oh man, we shouldn't discuss in telegram in the first place.

It is the best idea i have now to unblock 0-orch team

**Comment by khaledkbadr** [comment]
@Dinaamagdy tested the `zeroctl copy` vdisk from the `nbd-eol-experiment` on packet and got this error

```ct=2017-09-10T08:03:37+0000 lvl=info msg="parsing positional arguments..." module=global caller=vdisk.go:44
t=2017-09-10T08:03:37+0000 lvl=info msg="generating tlog data for target vdisk `vdisktest`" module=global caller=vdisk.go:104
Error: failed to generate tlog data for vdisk `vdisktest` : ListBlockIndices failed for vdisk `template:ubuntu-1604`: ERR unknown command:evalsha
Usage:
  zeroctl copy vdisk source_vdiskid target_vdiskid [target_clusterid] [flags]

Flags:
      --config SourceConfig   config resource: dialstrings (etcd cluster) or path (yaml file) (default config.yml)
      --data-shards int       data shards (K) variable of erasure encoding (default 4)
  -h, --help                  help for vdisk
      --parity-shards int     parity shards (M) variable of erasure encoding (default 2)
      --priv-key string       private key (default "12345678901234567890123456789012")
      --same                  enable flag to force copy within the same nbd servers

Global Flags:
  -v, --verbose   log available information

failed to generate tlog data for vdisk `vdisktest` : ListBlockIndices failed for vdisk `template:ubuntu-1604`: ERR unknown command:evalsha```

**Comment by grimpy** [comment]
When using hub.gig.tech as templatecluster we get the above error, this is because hub.gig.tech blocks all commands that can do a potential write operation.

**Comment by iwanbk** [comment]
> this is because hub.gig.tech blocks all commands that can do a potential write operation.

As far as i can see, there is no write operation in `ListBlockIndices` func.

But we are using Lua script there. I guess this is where the issue is, right?

**Comment by GlenDC** [comment]
> When using hub.gig.tech as templatecluster we get the above error, this is because hub.gig.tech blocks all commands that can do a potential write operation.

I get that @grimpy, but can that not be something blocked on a more fine-grained level. I mean in the end a Lua script is still only calling redis commands. So is there no way to block this more properly, it would really suck (for you) if we cannot use Lua scripts, which is what the `evalsha` command is for.

**Comment by grimpy** [comment]
@maxux would it be possible to open this? I mean withouth jeperdazing security, when allowing this command would the lua scripts be able to do write operation?

**Comment by GlenDC** [comment]
I'm not asking you to allow write operations @grimpy, as we do not need that ever. No 0-Disk process will ever write to the ETCD Config Cluster or template storage cluster.

I'm asking you to allow lua scripts which do not do any write operations. I think you understand that already, but I want to be clear as your request to @maxux is a bit ambiguous/confusing.

**Comment by maxux** [comment]
From what I see, ´evalsha` allows to run a "script" server side.
This is okay for me, as soon as this script cannot do more commands than a regular client (aka, no write operation). Can someone test that if a command is not allowed via a regular call, it's denied via the script too ?

**Comment by GlenDC** [comment]
How do you do it right now @maxux, I hope you are not manually keeping a witelist of commands you accept. Is there no configuration in the server (that you are using) that allows you to only accept readonly connections?

**Comment by maxux** [comment]
```
rename-command FLUSHDB ""
rename-command FLUSHALL ""
rename-command CONFIG ""
rename-command SLAVEOF ""
rename-command DEBUG ""
rename-command CLIENT ""
rename-command EVAL ""
rename-command EVALSHA ""
rename-command IMPORT ""
```

We deny explictly `evalsha`, so I guess it's "potentially" dangerous :/

**Comment by GlenDC** [comment]
It isn't. I mean it is just calling commands itself. You really should have some way that you accept read-only connections, which allows for any command and blocks write operations (ran directly or indirectly).

**Comment by maxux** [comment]
In theory, write are denied as soon as `SLAVEOF` is executed (which is the case).
Can somebody confirm that `write/set/any command` renamed are not available as well from the script ? I can't test right now :/

**Comment by GlenDC** [comment]
> Can somebody confirm...

Which server are you guys using? If you link me to the repo of that server technology I don't mind searching the answer for you in its codebase.

**Comment by maxux** [comment]
Upstream `ardb` with commands above on the config file

**Comment by GlenDC** [comment]
as in https://github.com/yinqiwen/ardb?

**Comment by maxux** [comment]
Yes

**Comment by iwanbk** [comment]
> Which server are you guys using? If you link me to the repo of that server technology I don't mind searching the answer for you in its codebase.

@GlenDC any update about this? Do we need to convert the Lua script to Go?

**Comment by iwanbk** [comment]
> In theory, write are denied as soon as SLAVEOF is executed (which is the case).
> Can somebody confirm that write/set/any command renamed are not available as well from the script ? I can't test right now :/

Just tested it.
If my below tests is the right procedure, slaveof will also prevent write operation from the script.

start two ardb servers at port 16386 and 16387

set the 16387 as slave of 16386
```
127.0.0.1:16387> slaveof 127.0.0.1 16386
OK
```
try `set` operation to 16387, it failed as expected
```
127.0.0.1:16387> set a 1
(error) READONLY You can't write against a read only slave.
````

try set operation using lua script, it failed
```
127.0.0.1:16387> eval "return redis.call('set','foo','bar')" 0
(error) ERR @user_script: 1: -READONLY You can't write against a read only slave.
```


disable slave
```
127.0.0.1:16387> slaveof no one
[9045] 09-11 16:51:10,794 INFO [Slave]Replication connection closed at state:INVALID_STATE
[ardb] 0:redis-cli*Z                                                                      [9045] 09-11 16:51:10,794 WARN Remove snapshot file:/tmp/tlog1/backup/sync-lmdb-backup.0.0.1505123164              "siputih" 16:51 11-Sep-17
                                                                                                                                                                                               [9044] 09-11 16:51:10,794 WARN Slave ipOK27.0.0.1,port=16387 closed.
```

set using lua script
```
127.0.0.1:16387> eval "return redis.call('set','foo','bar')" 0
OK
```

**Comment by maxux** [comment]
Okay, it's fine to open `EVALSHA` on the hub for me :)

**Comment by khaledkbadr** [comment]
@maxux Can you notify me when `EVALSHA` is open in the hub so we can test?

**Comment by GlenDC** [comment]
@iwanbk you don't seem to test if a `read-only` command from a lua script does work on a read-only slave.

**Comment by khaledkbadr** [comment]
As QA, @Dinaamagdy, just verified that the rollback finally worked on packet with a reasonable time for the create and the rollback.

**Comment by GlenDC** [comment]
Fixed by merging PR #453 into [the 1.1.0-alpha-8 branch](https://github.com/zero-os/0-Disk/tree/1.1.0-alpha-8), which will be used to make the official alpha-8 release as soon as the beta-1 milestone has come to an end.
